File: src/data/courses/dsa/trees/traversal-block/morris-traversal/code/java/MorrisTraversal.java
// ============================================================
// Morris Inorder Traversal
// ============================================================
import java.util.*;
class Solution {
    public List<Integer> morrisInorder(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        TreeNode cur = root;
        while (cur != null) {
            if (cur.left == null) {
                result.add(cur.val); cur = cur.right;
            } else {
                TreeNode pred = cur.left;
                while (pred.right != null && pred.right != cur) pred = pred.right;
                if (pred.right == null) { pred.right = cur; cur = cur.left; }
                else { pred.right = null; result.add(cur.val); cur = cur.right; }
            }
        }
        return result;
    }
}

File: src/data/courses/dsa/trees/traversal-block/morris-traversal/code/python/morris_traversal.py
# ============================================================
# Morris Inorder Traversal
# ============================================================
class Solution:
    def morrisInorder(self, root):
        result, current = [], root
        while current:
            if not current.left:
                result.append(current.val)
                current = current.right
            else:
                pred = current.left
                while pred.right and pred.right != current:
                    pred = pred.right
                if not pred.right:
                    pred.right = current  # Create thread
                    current = current.left
                else:
                    pred.right = None  # Remove thread
                    result.append(current.val)
                    current = current.right
        return result
